firedrake/preconditioners/pmg.py

```diff
@@ -274,55 +303,250 @@ def __init__(self, Vf, Vc, Vf_bcs, Vc_bcs):
+        # We transpose before and after the multiplcation times J to have each component
+        # stored contiguously as a scalar field, thus reducing the number of dgemm calls.
 
-void restriction(double *restrict Rc, const double *restrict Rf, const double *restrict w)
-{{
-    double Afc[{dimf}*{dimc}] = {{0}};
-    expression_kernel(Afc);
-    for (int32_t i = 0; i < {dimf}; i++)
-       for (int32_t j = 0; j < {dimc}; j++)
-           Rc[j] += Afc[i*{dimc} + j] * Rf[i] / w[i];
-}}
-"""
+        prolong_code = f"""
+        {kronmxv_code}
 
-        self.restrict_kernel = op2.Kernel(self.restrict_code, "restriction")
-        self.mesh = Vf.mesh()
+        void prolongation(double *restrict y, const double *restrict x){{
+            double JX[{mx}*{nx}] = {{ {JX} }};
+            double t0[{lwork}], t1[{lwork}];
+            double one=1.0E0;
+
+            for(int j=0; j<{nxyz}; j++)
+                for(int i=0; i<{nscal}; i++)
+                    t0[j + {nxyz}*i] = x[i + {nscal}*j];
+
+            kronmxv(0, {mx},{my},{mz}, {nx},{ny},{nz}, {nscal}, JX,JX,{JZ}, t0,t1);
+
+            for(int j=0; j<{mxyz}; j++)
+                for(int i=0; i<{nscal}; i++)
+                   y[i + {nscal}*j] = t1[j + {mxyz}*i];
+            return;
+        }}
+        """
+
+        restrict_code = f"""
+        {kronmxv_code}
 
+        void restriction(double *restrict y, const double *restrict x,
+        const double *restrict w){{
+            double JX[{mx}*{nx}] = {{ {JX} }};
+            double t0[{lwork}], t1[{lwork}];
+            double one=1.0E0;
+
+            for(int j=0; j<{mxyz}; j++)
+                for(int i=0; i<{nscal}; i++)
+                    t0[j + {mxyz}*i] = x[i + {nscal}*j] * w[i + {nscal}*j];
+
+            kronmxv(1, {nx},{ny},{nz}, {mx},{my},{mz}, {nscal}, JX,JX,{JZ}, t0,t1);
+
+            for(int j=0; j<{nxyz}; j++)
+                for(int i=0; i<{nscal}; i++)
+                    y[i + {nscal}*j] += t1[j + {nxyz}*i];
+            return;
+        }}
+        """
+
+        op2.configuration["ldflags"] = "-lblas"
+        self.prolong_kernel = op2.Kernel(prolong_code, "prolongation")
+        self.restrict_kernel = op2.Kernel(restrict_code, "restriction")
+        return
+
+    @staticmethod
+    def get_nodes_1d(V):
+        # Return GLL nodes if V==CG or GL nodes if V==DG
+        from FIAT import quadrature
+        from FIAT.reference_element import DefaultLine
+        issupported, N, family = tensor_product_space_query(V)
+        assert issupported
+        if set(family).issubset(["Q", "CG", "Lagrange"]):
+            rule = quadrature.GaussLobattoLegendreQuadratureLineRule(DefaultLine(), N+1)
+        elif set(family).issubset(["DQ", "DG", "Discontinuous Lagrange"]):
+            rule = quadrature.GaussLegendreQuadratureLineRule(DefaultLine(), N+1)
+        else:
+            raise ValueError("Don't know how to get nodes for %r" % family)
+
+        return np.matrix.flatten(rule.get_points())
```

Comment:
Don't use `np.matrix` (it's been deprecated for about 10 years), use `np.asarray(rule.get_points()).flatten()`